### PR TITLE
U4-11353 doctype specific tours

### PR DESCRIPTION
**Pull request by dawoe**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have linked this PR to an issue on the tracker at http://issues.umbraco.org/issue/U4-11353

### Description

This PR makes it possible to have tour for a specific doctype. The tour won't show up in the help drawer. On content items of the doctype a help button will be shown that starts a tour

### CODE REVIEW

**Comment by dawoe** [comment]
I just reviewed my own code... what i'm not happy with at the moment is using the Tour Api controller in the mapper. I think we need to introduce some kind of TourService. This will make it easier for developers to use the tours functionality in their own implementation

@Shazwazza and @zpqrtbnk ? What do you think ? I can make a new PR based on this one where I refactor some stuff ?

Dave

**Comment by dawoe** [comment]
Updated this PR so that a tour can be used for multiple doctypes. The contentType property of the tour now supports a comma seperated list of doctype aliases

**Comment by dawoe** [comment]
@nul800sebastiaan 

This PR works but I still think the code smells, because I reuse the Controller in the mapping. Could you ask @Shazwazza or @zpqrtbnk to look at this. I think adding a "TourService" will be a better solution.

Dave

File: src/Umbraco.Web/Models/BackOfficeTour.cs [hunk 1]
```diff
@@ -31,5 +31,8 @@ public BackOfficeTour()
 
         [DataMember(Name = "culture")]
         public string Culture { get; set; }
+
+        [DataMember(Name = "contentType")]
```
[the remark this answers is not in the document]
Comment (dawoe) [reply src/Umbraco.Web/Models/BackOfficeTour.cs]: This is used to specify the tour is for a doctype or multiple doctypes. When this is set with the alias, or comma seperated list of doctype aliases, the help button will be shown on the doctype. This is also used to filter out this tour in the help drawer

File: src/Umbraco.Web/Models/ContentEditing/ContentItemDisplay.cs [hunk 2]
```diff
@@ -57,5 +57,8 @@ public ContentItemDisplay()
 
         [DataMember(Name = "isBlueprint")]
         public bool IsBlueprint { get; set; }
+
+        [DataMember(Name = "tour")]
+        public string Tour { get; set; }
```
[the remark this answers is not in the document]
Comment (dawoe) [reply src/Umbraco.Web/Models/ContentEditing/ContentItemDisplay.cs]: I did this here so we can already retreive the associated tour for this doctype when getting the content. This avoids an api call when the content item is displayed to check if there is a tour associated with it

**Comment by dawoe** [comment]
Hi @Shazwazza 

Will wait for feedback of @madsrasmussen about the UI part. Will try to refactor it after that.

Dave

**Comment by dawoe** [comment]
Hi @Shazwazza 

I refactored it after your comments. Can you please have a look ?

@madsrasmussen can you have a look at UI part >

**Comment by madsrasmussen** [comment]
Hi Dave, 

I am not a big fan of the help button at the bottom. We want to keep all buttons in the footer related to the publishing of content.

The convention we currently have for help in the back office is the help drawer. It is able to show context relevant help and we already do this for many sections. Right now that is the correct place to add the doctype tours until we find a more general convention on how to handle inline help in the back office. One way to do this is to add a new section at the top which will be shown for doctype tours called something like: "Need help to enter content for (Doc type name)".

**Comment by dawoe** [comment]
Hi @madsrasmussen 

Thanks for the feedback. Maybe I should clarify why I made this PR.

For a customer we implemented tours. But they had a very complex user model set up. They have about 130 low frequency editors. A lot of them having different start nodes and even different content structures in their part of the site. Starting a tour to explain how a doctype works was hard. Because the tour needs to start by having them go to a item of that doctype. Seeing they have different start nodes and different content structures (but can be based on the same doctypes) this was impossible to do.

We hacked around a little bit and came up with a solution like this help button. Although I agree the placement of the button is not that ideal.

So with this PR my main goal was to have tours that don't show up in the help drawer but enable a help button on a item of a specific doctype. So the tour is confined to the currently edited item.

So in my opinion moving this to the help drawer buries the help for the doctype you currently working on to deep for most low frequency editors. I don't know if the help drawer is context aware enough to know which item you are editing ???

Maybe moving the button to header and make it only show the help icon would be better. So not really a button if you get what I mean. Will try to make a mockup.

Dave

**Comment by madsrasmussen** [comment]
I do understand that things sometimes can disappear in the help drawer but that is another task to solve. 

The help drawer should be able to get the currently open node with something like editorState service.

If you need a button to be on the node you can always build a custom property editor and use that to launch the tour. I think there are a couple of packages which already adds the ability to show inline help for a document type. It might make sense to team up with them and include this functionality too.

**Comment by dawoe** [comment]
Will try to have a stab integrating it into the help drawer. Probably somewhere over the weekend

Dave

**Comment by dawoe** [comment]
Hi @madsrasmussen 

Did some updates to this PR. I described the changes on the issue on the tracker. I also uploaded a tour file and video there demonstrating the new functionality.

What do you think ?

Dave

**Comment by dawoe** [comment]
Hi @madsrasmussen, @nul800sebastiaan 

When updating my tour editor package for this functionality I discovered "doctype tours" where not displayed for items that are child from a list view node.

This was because I was showing the tours based on the tree state....which does not change when you select a child item. I solved it by emitting a event when the editorState changes and listen to do that.

Dave

**Comment by dawoe** [comment]
@nul800sebastiaan is this one still on the radar. I'm holding of a new release of my tour editor package because of this one. And I want to start on a V8 compatible version of the tour editor

**Comment by nul800sebastiaan** [comment]
Hey there @dawoe  !

I'm writing today after announcing at Codegarden, our yearly conference, that the upcoming version 7.15.0 will be the last Umbraco v7 release with new features in it. From now on all our efforts will be focused on version 8 instead. 🎱

We're wrapping up the 7.15 version at the moment and unfortunately these proposed changes won't be able to make it in there.

We'd like to extend a big thank you the efforts on this so far and we would love it if you could have a look at porting this feature over to v8 instead. Of course if you have any questions on where to start then feel free to ask on the forum, we monitor all questions there and we're happy to assist.

Thanks again for the work in this PR and we hope to see you contributing to v8 some time soon!

Best, Sebastiaan and the Umbraco CMS team.

**Pull request closed without merging**